Add gas free EVM gateway  guide (#1291)
onflow/docs: docs/tutorials/backend-usage/gas-free-evm-endpoint.md

So here is the problem - if we want to send a EVM transaction through the default EVM Gateway,  the sender's account has enough balance to cover the transaction fees. However, in some scenarios, developers may prefer to sponsor the transaction fees to create a more user-friendly experience, just can be done in native Cadence transactions, or paymaster solutions on other networks.

Here we provide a solution - you can set up a sponsored transaction EVM endpoint for your backend service and all transactions sent through this endpoint will be sponsored by your account.

### Who needs the gas free EVM endpoint?

Here are several typical business scenarios suitable for its use:

- Centralized exchanges that wish to improve internal transaction performance for users' deposits and withdrawals.
- Apps with embedded wallets using social login that don't want to require users to deposit money in a new wallet before they're able to interact with the app.

> - Games needing to complete onchain tasks to onboard new players who wish to try the game.
> - Businesses that wish to provide sponsored transactions as a service.
> 

Reviewer: ```suggestion
- Games needing to complete onchain tasks to onboard new players who wish to try the game.
- Backends creating and managing large amounts of wallets on behalf of their users.
- Businesses that wish to provide sponsored transactions as a service.

```

---
title: Sponsor Transactions RPC Node

> description: Learn how to set up an EVM RPC endpoint for your app and use it to sponsor all transactions for your users.

Reviewer: ```suggestion
description: Learn how to set up an EVM RPC endpoint for your app and use it to sponsor all transactions sent to it.
```

Here we provide a solution - you can set up a sponsored transaction EVM endpoint for your backend service and all transactions sent through this endpoint will be sponsored by your account.

### Who needs the gas free EVM endpoint?

Here are several typical business scenarios suitable for its use:

- Apps with embedded wallets using social login that don't want to require users to deposit money in a new wallet before they're able to interact with the app.
- Games needing to complete onchain tasks to onboard new players who wish to try the game.
- Centralized or decentralized exchanges that wish to improve user experience for a competitive edge
- Businesses that wish to provide sponsored transactions as a service.

As long as you can ensure that all your users' transactions are sent through a custom RPC endpoint, you can set up a Gas Free EVM Gateway to enable gas-free transactions for your users.

## How to set up a gas free EVM endpoint?

Mosts of the tasks you need to complete are in the guide to set up your own [Custom EVM Gateway].  A few highlights:

> 1. You need to prepare a Service Account for the EVM Gateway to cover all transaction fees.
> 
> Please refer to [Account Creation] for more details.

Reviewer: ```suggestion
1. You need to prepare a Service Account for the EVM Gateway to cover all transaction fees. Please refer to [Account Creation] for more details.
```

2. You need to add enough identical keys to the Service Account to support the concurrent signing of EVM transactions. This is very important, because the EVM Gateway will use the Service Account to pay for the gas fees of the EVM transactions, so the Service Account needs to have enough keys to support the concurrent signing of EVM transactions. Please refer to [Account and Key Management] for more details.

3. You need to set the correct environment variables for the EVM Gateway to enable the gas-free feature, adjust the environment variables in the [Run the gateway] section as follows:

- `COINBASE`: The address is used to accept EVM transaction fees.  In this case, there won't be fees to accept because all fees will be covered by the service account. But you need to set it with a valid address to ensure the EVM Gateway can start.
- `COA_ADDRESS`: This is the service account address, which will be used to pay for the gas fees of the EVM transactions. Please input the address of the Service Account you created in step 1, but without the `0x` prefix.

>   - **Fund this address** to cover transaction fees.
>   - **Fund this address** to cover transaction fees.

Reviewer: ```suggestion
  - **Fund this address** to cover transaction fees.
```

### Why we need a gas free EVM endpoint?

From the Flow [transaction model], we know that there is actually a Fee Payer role in native Flow transactions. When the transaction is executed, the fees for the transaction are entirely borne by the Fee Payer role.  However, for EVM, the transaction model doesn't separate the signer from the payer. Therefore, when we execute transactions on the EVM, the transaction fees **must** be covered by the sender of the transaction.

So here is the problem - if we want to send a EVM transaction through the default EVM Gateway,  the sender's account has enough balance to cover the transaction fees. However, in some scenarios, developers may prefer to sponsor the transaction fees to create a more user-friendly experience, just can be done in native Cadence transactions, or paymaster solutions on other networks.

Here we provide a solution - you can set up a sponsored transaction EVM endpoint for your backend service and all transactions sent through this endpoint will be sponsored by your account.

### Who needs the gas free EVM endpoint?

Here are several typical business scenarios suitable for its use:

> - Centralized Exchange or any EVM Wallets Custodial service provider on Flow EVM. They usually need to support users to deposit, withdraw and manage EVM assets, and they usually need to cover the transaction fees for users in a more convenient way.
> - For the App with cutodial wallet integation which can use custom RPC endpoint set by App builder. They can ensure their users' transactions are sent through their own specified RPC endpoint.
> 
> As long as you can ensure that all your users' transactions are sent through a custom RPC endpoint, you can set up a Gas Free EVM Gateway to enable gas-free transactions for your users.

Reviewer: ```suggestion
- Apps with embedded wallets using social login that don't want to require users to deposit money in a new wallet before they're able to interact with the app.
- Games needing to complete onchain tasks to onboard new players who wish to try the game.
- Centralized or decentralized exchanges that wish to improve user experience for a competitive edge
- Businesses that wish to provide sponsored transactions as a service.

As long as you can ensure that all your users' transactions are sent through a custom RPC endpoint, you can set up a Gas Free EVM Gateway to enable gas-free transactions for your users.
```

## Context

In [Using Flow EVM], we know that we can use the RPC endpoint to send EVM transactions to the Flow EVM. The default RPC endpoint is actually the [EVM Gateway].

### How does the EVM Gateway work?

For any EVM transactions sent through the EVM Gateway, Essentially, they will be wrapped into a native Flow Cadence transaction and sent to the Flow network and Flow EVM Gateway's service account is the payer of that Cadence transaction. But for EVM transactions inside the native transaction, the transaction fees of the EVM transactions will be extracted from the sender's account and charged to the EVM Gateway's service account.

Here is the key points of the Flow EVM Gateway:

- The EVM Gateway works as a proxy to send the EVM transactions.
- The EVM Gateway wraps the EVM transactions into native Flow Cadence transactions.
- The EVM Gateway has a service account, which is the payer and sender of all native transactions sent through the EVM Gateway.
- The sender of the EVM transaction is the fee payer of its EVM transaction to the EVM Gateway's service account.

### Why we need a gas free EVM endpoint?

> From the Flow [transaction model], we know that there is actually a Fee Payer role in native Flow transactions. When the transaction is executed, the fees for the transaction are entirely borne by the Fee Payer role.  
> However, for Flow EVM, the concept of a Fee Payer does not exist in the EVM transaction model. Therefore, when we execute transactions on the EVM, the transaction fees need to be covered by the sender of the transaction.

Reviewer: ```suggestion
From the Flow [transaction model], we know that there is actually a Fee Payer role in native Flow transactions. When the transaction is executed, the fees for the transaction are entirely borne by the Fee Payer role.  However, for EVM, the transaction model doesn't separate the signer from the payer. Therefore, when we execute transactions on the EVM, the transaction fees **must** be covered by the sender of the transaction.
```

So here is the problem - if we want to send a EVM transaction through the default EVM Gateway,  the sender's account has enough balance to cover the transaction fees. However, in some scenarios, developers may prefer to sponsor the transaction fees to create a more user-friendly experience, just can be done in native Cadence transactions, or paymaster solutions on other networks.

Here we provide a solution - you can set up a sponsored transaction EVM endpoint for your backend service and all transactions sent through this endpoint will be sponsored by your account.

### Who needs the gas free EVM endpoint?

> Here are several typical business scenarios suitable for its use.

Reviewer: ```suggestion
Here are several typical business scenarios suitable for its use:
```

---

> title: Gas Free EVM Endpoint

Reviewer: ```suggestion
title: Sponsor Transactions RPC Node
```

> description: Learn how to set up a gas free EVM endpoint for your backend, all transactions sent through this endpoint will not be charged for gas fees from the transaction sender's account(EVM Gateway's service account will be used to pay for the gas fees).

Reviewer: ```suggestion
description: Learn how to set up an EVM RPC endpoint for your app and use it to sponsor all transactions for your users.
```

sidebar_position: 2
keywords:
  - Backend
  - Flow
  - Flow EVM
  - Flow Node
  - Gas Free
  - EVM Gateway
  - RPC Endpoint
---

# Gas Free EVM Endpoint

In this tutorial, we will learn how to set up a gas free EVM endpoint for your backend service, all transactions sent through this endpoint will not be charged for gas fees from the transaction sender's account.

## Context

In [Using Flow EVM], we know that we can use the RPC endpoint to send EVM transactions to the Flow EVM. The default RPC endpoint is actually the [EVM Gateway].

### How does the EVM Gateway work?

> For any EVM transactions sent through the EVM Gateway, Essentially, they will be wrapped into a native Flow Cadence transaction and sent to the Flow network and Flow EVM Gateway's service account is the payer of that Cadence transaction. But for EVM transactions inside the native transaction, the transaction fees of the EVM transactions will be extracted from the sender's account and charged to the EVM Gateway's service account.

Reviewer: ```suggestion
EVM transactions sent through the EVM Gateway are wrapped in a Flow Cadence transaction and sent to the Flow network. The Flow EVM Gateway's service account is the payer of that Cadence transaction.

:::info

In Cadence, the proposer, payer, and signer of a transaction are separate, natively.

:::

For EVM transactions inside this Cadence transaction, the transaction fees of the EVM transaction sent from the sender's account and applied to the EVM Gateway's service account.  The net result is that the EVM Gateway gets refunded for the same amount it spent on the Cadence transaction.
```

Reviewer: @btspoony is this still accurate?

Reviewer: Chef says it is, please let me know if that's not true.

Here is the key points of the Flow EVM Gateway:

> - The EVM Gateway is working as a proxy to send the EVM transactions.

Reviewer: ```suggestion
- The EVM Gateway works as a proxy to send the EVM transactions.
```

> - The EVM Gateway will wrap the EVM transactions into native Flow Cadence transactions.

Reviewer: ```suggestion
- The EVM Gateway wraps the EVM transactions into native Flow Cadence transactions.
```

- The EVM Gateway has a service account, which is the payer and sender of all native transactions sent through the EVM Gateway.

> - The sender of the EVM transaction will be the fee payer of its EVM transaction to the EVM Gateway's service account.

Reviewer: ```suggestion
- The sender of the EVM transaction is the fee payer of its EVM transaction to the EVM Gateway's service account.
```

### Why we need a gas free EVM endpoint?

> From Flow's [transaction model], we know that there is actually a Fee Payer role in native Flow transactions. When the transaction is executed, the fees for the transaction are entirely borne by the Fee Payer role.  

Reviewer: ```suggestion
From the Flow [transaction model], we know that there is actually a Fee Payer role in native Flow transactions. When the transaction is executed, the fees for the transaction are entirely borne by the Fee Payer role.  
```

However, for Flow EVM, the concept of a Fee Payer does not exist in the EVM transaction model. Therefore, when we execute transactions on the EVM, the transaction fees need to be covered by the sender of the transaction.

> So here is the problem - if we want to send a EVM transaction through the default EVM Gateway, we need to make sure the sender's account has enough balance to cover the transaction fees.  
> However, in some scenarios, developers may prefer a more user-friendly experience, wishing that EVM transactions could also support the ability for a Fee Payer to fully cover the transaction fees, just like the native Cadence transactions do.
> 
> Here we provide a solution, you can set up a gas free EVM endpoint for your backend service, all transactions sent through this endpoint will not be charged for gas fees from the transaction sender's account.

Reviewer: ```suggestion
So here is the problem - if we want to send a EVM transaction through the default EVM Gateway,  the sender's account has enough balance to cover the transaction fees. However, in some scenarios, developers may prefer to sponsor the transaction fees to create a more user-friendly experience, just can be done in native Cadence transactions, or paymaster solutions on other networks.

Here we provide a solution - you can set up a sponsored transaction EVM endpoint for your backend service and all transactions sent through this endpoint will be sponsored by your account.
```

Here are several typical business scenarios suitable for its use:

- Apps with embedded wallets using social login that don't want to require users to deposit money in a new wallet before they're able to interact with the app.
- Games needing to complete onchain tasks to onboard new players who wish to try the game.
- Centralized or decentralized exchanges that wish to improve user experience for a competitive edge
- Businesses that wish to provide sponsored transactions as a service.

As long as you can ensure that all your users' transactions are sent through a custom RPC endpoint, you can set up a Gas Free EVM Gateway to enable gas-free transactions for your users.

## How to set up a gas free EVM endpoint?

> Here is the guide to set up a [Custom EVM Gateway], but that's the general guide, we will focus on the gas-free part in this tutorial.

Reviewer: ```suggestion
Mosts of the tasks you need to complete are in the guide to set up your own [Custom EVM Gateway].  Additionally, complete the tasks below:
```

> Mosts of the tasks you need to complete are in the guide to set up your own [Custom EVM Gateway].  Additionally, complete the tasks below:

Reviewer: ```suggestion
Mosts of the tasks you need to complete are in the guide to set up your own [Custom EVM Gateway].  A few highlights:
```

- Centralized or decentralized exchanges that wish to improve user experience for a competitive edge
- Businesses that wish to provide sponsored transactions as a service.

As long as you can ensure that all your users' transactions are sent through a custom RPC endpoint, you can set up a Gas Free EVM Gateway to enable gas-free transactions for your users.

## How to set up a gas free EVM endpoint?

Mosts of the tasks you need to complete are in the guide to set up your own [Custom EVM Gateway].  A few highlights:

1. You need to prepare a Service Account for the EVM Gateway to cover all transaction fees. Please refer to [Account Creation] for more details.

2. You need to add enough identical keys to the Service Account to support the concurrent signing of EVM transactions. This is very important, because the EVM Gateway will use the Service Account to pay for the gas fees of the EVM transactions, so the Service Account needs to have enough keys to support the concurrent signing of EVM transactions. Please refer to [Account and Key Management] for more details.

> 3. You need to set the correct environment variables for the EVM Gateway to enable the gas-free feature, adjust the environment variables in the [Run the gateway] section as follows:
> 
> - `COINBASE`: The address is used to accept EVM transaction fees, but in this case, it cannot to accept fees because all fees will be covered by the service account. But you need to set it with a valid address to ensure the EVM Gateway can start.

Reviewer: ```suggestion
3. You need to set the correct environment variables for the EVM Gateway to enable the gas-free feature, adjust the environment variables in the [Run the gateway] section as follows:

- `COINBASE`: The address is used to accept EVM transaction fees.  In this case, there won't be fees to accept because all fees will be covered by the service account. But you need to set it with a valid address to ensure the EVM Gateway can start.
```

> - `COA_ADDRESS`: This is the service account address, which will be used to pay for the gas fees of the EVM transactions. Please input the address of the Service Account you created in step 1, but without the `0x` prefix.

Reviewer: ```suggestion
- `COA_ADDRESS`: This is the service account address, which will be used to pay for the gas fees of the EVM transactions. Please input the address of the Service Account you created in step 1, but without the `0x` prefix.
  - **Fund this address** to cover transaction fees.
```

- `COA_KEY`: You need to set the private key of the Service Account you created in step 1.

> - `GAS_PRICE`: That is the most important part, you need to set it to `0` to ensure the linked service account will pay for transactions on users behalf.

Reviewer: ```suggestion
- `GAS_PRICE`: **Critical**: set this to `0` to ensure the linked service account will pay for transactions on users behalf.
```

> 4. Please follow the full guide of [Custom EVM Gateway] to complete the EVM gateway setup, then you will get a custom RPC endpoint which is 100% gas-free for any EVM transaction sent through it.

Reviewer: ```suggestion
4. Please follow the full guide of [Custom EVM Gateway] to complete the EVM gateway setup, then you will get a custom RPC endpoint which will sponsor 100% of the gas fees for any EVM transaction sent through it.
```

---
title: Sponsor Transactions RPC Node
description: Learn how to set up an EVM RPC endpoint for your app and use it to sponsor all transactions for your users.
sidebar_position: 2
keywords:
  - Backend
  - Flow
  - Flow EVM
  - Flow Node
  - Gas Free
  - EVM Gateway
  - RPC Endpoint
---

# Gas Free EVM Endpoint

In this tutorial, we will learn how to set up a gas free EVM endpoint for your backend service, all transactions sent through this endpoint will not be charged for gas fees from the transaction sender's account.

## Context

In [Using Flow EVM], we know that we can use the RPC endpoint to send EVM transactions to the Flow EVM. The default RPC endpoint is actually the [EVM Gateway].

### How does the EVM Gateway work?

EVM transactions sent through the EVM Gateway are wrapped in a Flow Cadence transaction and sent to the Flow network. The Flow EVM Gateway's service account is the payer of that Cadence transaction.

:::info

In Cadence, the proposer, payer, and signer of a transaction are separate, natively.

:::

> For EVM transactions inside this Cadence transaction, the transaction fees of the EVM transaction sent from the sender's account and applied to the EVM Gateway's service account.  The EVM Gateway pays the Cadence transaction fee, but is reimbursed via the EVM transaction's gas fee, which is directed to its own EVM address as the "coinbase".

Reviewer: ```suggestion
For EVM transactions inside this Cadence transaction, the transaction fees of the EVM transaction sent from the sender's account and applied to the EVM Gateway's service account.  The EVM Gateway pays the Cadence transaction fee, but is reimbursed via the EVM transaction's gas fee, which is directed to its own EVM address as the _coinbase_.
```

- Games needing to complete onchain tasks to onboard new players who wish to try the game.
- Centralized or decentralized exchanges that wish to improve user experience for a competitive edge
- Businesses that wish to provide sponsored transactions as a service.

As long as you can ensure that all your users' transactions are sent through a custom RPC endpoint, you can set up a Gas Free EVM Gateway to enable gas-free transactions for your users.

## How to set up a gas free EVM endpoint?

Mosts of the tasks you need to complete are in the guide to set up your own [Custom EVM Gateway].  A few highlights:

1. You need to prepare a Service Account for the EVM Gateway to cover all transaction fees. Please refer to [Account Creation] for more details.

2. You need to add enough identical keys to the Service Account to support the concurrent signing of EVM transactions. This is very important, because the EVM Gateway will use the Service Account to pay for the gas fees of the EVM transactions, so the Service Account needs to have enough keys to support the concurrent signing of EVM transactions. Please refer to [Account and Key Management] for more details.

> 3. You need to set the correct environment variables for the EVM Gateway to archive the gas-free feature.

Reviewer: ```suggestion
3. You need to set the correct environment variables for the EVM Gateway to enable the gas-free feature:
```

> 3. You need to set the correct environment variables for the EVM Gateway to enable the gas-free feature:
> 
> In the [Run the gateway] section, you need to set couple of environment variables, and here are the notes:

Reviewer: ```suggestion
3. You need to set the correct environment variables for the EVM Gateway to enable the gas-free feature, adjust the environment variables in the [Run the gateway] section as follows:
```
